web3/contract/utils.py

```diff
@@ -365,6 +373,46 @@ def get_function_by_identifier(
     return fns[0]
 
 
+def find_events_by_identifier(
+    contract_abi: ABI,
+    w3: Union["Web3", "AsyncWeb3"],
+    address: ChecksumAddress,
+    callable_check: Callable[..., Any],
+    event_type: Type[TContractEvent],
+) -> List[TContractEvent]:
+    """
+    Given a contract ABI, return a list of TContractEvent instances.
+    """
+    event_abis = filter_abi_by_type("event", contract_abi)
+    return [
+        event_type.factory(
+            event_abi["name"],
+            w3=w3,
+            contract_abi=contract_abi,
+            address=address,
+            abi=event_abi,
+        )
+        for event_abi in event_abis
+        if callable_check(event_abi)
+    ]
+
+
+def get_event_by_identifier(
+    events: Sequence[TContractEvent], identifier: str
+) -> TContractEvent:
+    """
+    Check that the provided list of TContractEvent instances contains one element and
+    return it.
+    """
+    if len(events) > 1:
+        raise Web3ValueError(
+            f"Found multiple events with matching {identifier}. " f"Found: {events!r}"
```

Comment:
nit: use "identifier" in the message for clarity 

```suggestion
            f"Found multiple events with matching identifier `{identifier}`: {events!r}"
```

Reply:
I think `identifier` is a little confusing because we use `abi_element_identifier` to represent the name of the `ABIFunction` or `ABIEvent`.

In this context, `identifier` is a string that represents the type of "check" passed in. For example, `get_function_by_signature` uses `find_functions_by_identifier` by passing a "check" function that retrieves all functions from the ABI which passes the check. Then the resulting list of ABI components is passed to `get_function_by_identifier` to ensure there is only one result in the list and returns the ABI component itself. If not, an exception is raised, and the `identifier` is `"signature"`, indicating the search type.